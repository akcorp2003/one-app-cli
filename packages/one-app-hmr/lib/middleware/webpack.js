/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';
import webpackDevMiddleware from 'webpack-dev-middleware';
import webpackHotMiddleware from 'webpack-hot-middleware';

import { createHotModuleWebpackConfig, buildExternalsDLL } from '../webpack';
import {
  info, error, warn, time, log, yellow, orange, dodgerblue,
} from '../logs';
import { vfs } from '../utils';

export const printWebpack = (message) => `${dodgerblue('webpack')} - ${message}`;

export function printStatsWhenDone(stats) {
  const { errors, warnings } = stats.compilation;
  errors.forEach((message) => error(message));
  warnings.forEach((message) => warn(message));
  log(printWebpack(`webpack built in ${orange(`${stats.endTime - stats.startTime}`)} ms`));
}

export function printWhenInvalid() {
  warn(printWebpack('webpack building...'));
}

export async function loadWebpackMiddleware({
  context,
  publicPath,
  staticPath,
  modules,
  externals,
  rootModuleName,
  serverAddress,
} = {}) {
  log(printWebpack('initializing webpack'));

  // TODO: only run if externals present
  await time(printWebpack(orange('pre-building dll externals for Holocron modules')), async () => {
    await buildExternalsDLL({ externals });
  });

  const webpackConfig = createHotModuleWebpackConfig({
    context,
    publicPath,
    staticPath,
    modules,
    externals,
    rootModuleName,
  });
  const compiler = webpack(webpackConfig);
  compiler.hooks.done.tap('OneAppHMR', printStatsWhenDone);
  compiler.hooks.invalid.tap('OneAppHMR', printWhenInvalid);
  // removing all logs
  // https://webpack.js.org/configuration/other-options/#infrastructurelogging
  const devMiddleware = webpackDevMiddleware(compiler, {
    stats: false,
    index: false,
    serverSideRender: true,
    writeToDisk: !true,
    outputFileSystem: vfs,
    publicPath: webpackConfig.output.publicPath,
  });
  const hotMiddleware = webpackHotMiddleware(compiler, {
    reload: true,
    log: false,
    dynamicPublicPath: false,
  });

  devMiddleware.waitUntilValid(() => {
    info(`${orange('ðŸ”¥ HMR server is ready')} - visit "${yellow(serverAddress)}" to start!\n`);
  });

  return {
    devMiddleware,
    hotMiddleware,
    publish: (...args) => hotMiddleware.publish(...args),
  };
}
