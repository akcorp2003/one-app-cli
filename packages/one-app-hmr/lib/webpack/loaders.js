/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import ExtractCssChunks from 'extract-css-chunks-webpack-plugin';

export const fileLoader = {
  test: /\.(woff|woff2|ttf|eot|svg|png|jpg|jpeg|gif|webm)(\?.*)?$/,
  use: [{
    loader: 'file-loader',
    options: {
      name: '../assets/[name].[ext]',
    },
  }],
};

export const cssLoader = {
  test: /\.(sa|sc|c)ss$/,
  exclude: /node_modules/,
  use: [
    {
      loader: ExtractCssChunks.loader,
      options: {
        hmr: true,
        esModule: true,
        modules: {
          namedExport: true,
        },
      },
    },
    {
      loader: 'css-loader',
      options: {
        importLoaders: 2,
        modules: {
          localIdentName: '[name]__[local]___[hash:base64:5]',
        },
      },
    },
    {
      loader: 'sass-loader',
    },
    {
      loader: 'postcss-loader',
      options: {
        sourceMap: true,
        postcssOptions: {
          syntax: 'postcss-scss',
          plugins: {
            'postcss-preset-env': {
              browsers: 'last 2 versions',
            },
            cssnano: {},
            'postcss-browser-reporter': {},
          },
        },
      },
    },
  ],
};

export const jsxLoader = ({ plugins = [], presets = [] } = {}) => ({
  test: /\.jsx?$/,
  use: [
    {
      loader: 'babel-loader',
      options: {
        cwd: path.resolve(__dirname, '..'),
        cacheDirectory: true,
        babelrc: false,
        // TODO: make extensible for special cases
        presets: [
          [
            'amex',
            {
              'preset-env': {
                modules: false,
              },
            },
          ],
          ...presets,
        ],
        plugins,
      },
    },
  ],
});
