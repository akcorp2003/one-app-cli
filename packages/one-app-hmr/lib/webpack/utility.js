/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import TerserPlugin from 'terser-webpack-plugin';

import { isDevelopment } from '../utils';

// eslint-disable-next-line camelcase, import/prefer-default-export
export function createMinifyConfig({ isDev = isDevelopment(), keep_fnames = true } = {}) {
  return {
    devtool: isDev && 'eval-cheap-source-map',
    optimization: {
      minimize: !isDev,
      minimizer: [
        new TerserPlugin({
          test: /\.jsx?$/i,
          terserOptions: {
            keep_fnames,
          },
        }),
      ],
    },
  };
}

export function createExternalEntry([packageName, varName]) {
  return {
    [packageName]: {
      commonjs2: packageName,
      ...varName ? {
        var: varName,
        root: varName,
      } : {},
    },
  };
}

export function createOneAppExternals(providedExternals = []) {
  return [
    ['@americanexpress/one-app-ducks', 'OneAppDucks'],
    ['@americanexpress/one-app-router', 'OneAppRouter'],
    ['create-shared-react-context', 'CreateSharedReactContext'],
    ['holocron', 'Holocron'],
    ['holocron-module-route', 'HolocronModuleRoute'],
    ['immutable', 'Immutable'],
    ['prop-types', 'PropTypes'],
    ['react', 'React'],
    ['react-dom', 'ReactDOM'],
    ['react-helmet', 'ReactHelmet'],
    ['react-redux', 'ReactRedux'],
    ['redux', 'Redux'],
    ['reselect', 'Reselect'],
    ...providedExternals.map((external) => (Array.isArray(external) ? external : [external])),
  ]
    .map(createExternalEntry)
    .reduce((map, next) => ({ ...map, ...next }), {});
}

export function createHotModuleEntry({ moduleName, modulePath }) {
  return {
    [moduleName]: [
      require.resolve('webpack-hot-middleware/client'),
      require.resolve('react-refresh/runtime'),
      `${modulePath}/src/index.js`,
    ],
  };
}

export function createHotModuleEntries(modules = []) {
  return modules
    .map(createHotModuleEntry)
    .reduce((map, next) => ({ ...map, ...next }), {});
}
