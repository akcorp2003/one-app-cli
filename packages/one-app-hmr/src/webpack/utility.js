/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';

import TerserPlugin from 'terser-webpack-plugin';

export const STATIC_DIR = 'static';
export const MODULES_DIR = 'modules';
export const EXTERNAL_DIR = 'vendor';
export const TEMP_DIR = '.temp';

export function isDevelopment() {
  return process.env.NODE_ENV === 'development';
}

export function getContext() {
  return process.cwd();
}

export function getStaticPath(filePath = '', context) {
  return path.resolve(context || getContext(), STATIC_DIR, filePath);
}

export function getModulesPath(filePath = '', context) {
  return path.resolve(context || getContext(), STATIC_DIR, MODULES_DIR, filePath);
}

export function getPublicPath(moduleName) {
  return [STATIC_DIR, MODULES_DIR].concat(moduleName || []).join('/');
}

export function getExternalsPublicPath(resource) {
  return [STATIC_DIR, EXTERNAL_DIR].concat(resource || []).join('/');
}

export function createExternalEntry([packageName, varName] = []) {
  return {
    [packageName]: {
      commonjs2: packageName,
      ...varName ? {
        var: varName,
        root: varName,
      } : {},
    },
  };
}

export function createOneAppExternals(providedExternals = []) {
  return [
    ['@americanexpress/one-app-ducks', 'OneAppDucks'],
    ['@americanexpress/one-app-router', 'OneAppRouter'],
    ['create-shared-react-context', 'CreateSharedReactContext'],
    ['holocron', 'Holocron'],
    ['holocron-module-route', 'HolocronModuleRoute'],
    ['immutable', 'Immutable'],
    ['prop-types', 'PropTypes'],
    ['react', 'React'],
    ['react-dom', 'ReactDOM'],
    ['react-helmet', 'ReactHelmet'],
    ['react-redux', 'ReactRedux'],
    ['redux', 'Redux'],
    ['reselect', 'Reselect'],
    ...providedExternals,
  ]
    .map(createExternalEntry)
    .reduce((map, next) => ({ ...map, ...next }), {});
}

export function createHotModuleEntry({ moduleName, modulePath } = {}) {
  return {
    [moduleName]: [
      'react-refresh/runtime',
      'webpack-hot-middleware/client',
      `${modulePath}/src/index.js`,
    ],
  };
}

export function createHotModuleEntries(modules = []) {
  return modules
    .map(createHotModuleEntry)
    .reduce((map, next) => ({ ...map, ...next }), {});
}

// eslint-disable-next-line camelcase
export function createMinifyConfig({ isDev = isDevelopment(), keep_fnames = true } = {}) {
  return {
    devtool: isDev && 'eval-cheap-source-map',
    optimization: {
      minimize: !isDev,
      minimizer: [
        new TerserPlugin({
          test: /\.jsx?$/i,
          terserOptions: {
            keep_fnames,
          },
        }),
      ],
    },
  };
}
