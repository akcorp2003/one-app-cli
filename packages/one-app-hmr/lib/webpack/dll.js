/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';
import merge from 'webpack-merge';

import {
  isDevelopment,
  getContextPath,
  getStaticPath,
} from '../utils';
import { createOneAppExternals, createMinifyConfig } from './utility';
import { jsxLoader } from './loaders';

// eslint-disable-next-line import/prefer-default-export
export function createDLLConfig({
  isDev = isDevelopment(),
  useAsReference = false,
  dllName = 'externals',
  manifestPathName = getStaticPath(`.${dllName}.dll.json`),
  entries = createOneAppExternals(),
  externals = [],
} = {}) {
  return {
    ...useAsReference
      ? {}
      : merge(
        createMinifyConfig({ isDev }),
        {
          mode: isDev ? 'development' : 'production',
          entry: { [dllName]: entries },
          externals,
          output: {
            path: getStaticPath(),
            filename: 'vendor/[name].js',
            library: dllName,
          },
          module: {
            rules: [jsxLoader()],
          },
        }
      ),
    plugins: useAsReference
      ? [
        new webpack.DllReferencePlugin({
          context: getContextPath(),
          name: dllName,
          manifest: manifestPathName,
        }),
      ]
      : [
        new webpack.DllPlugin({
          context: getContextPath(),
          name: dllName,
          path: manifestPathName,
        }),
      ],
  };
}
