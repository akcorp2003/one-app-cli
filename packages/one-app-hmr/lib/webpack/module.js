/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import webpack from 'webpack';
import merge from 'webpack-merge';
import ReactRefreshWebpackPlugin from '@pmmmwh/react-refresh-webpack-plugin';

import HotHolocronModulePlugin from './plugins/hot-holocron';

import {
  createOneAppExternals,
  createHotModuleEntries,
} from './utility';
import {
  jsxLoader,
  cssLoader,
  fileLoader,
} from './loaders';
import { createDLLConfig } from './dll';
import { getContextPath, getModulesPath, getPublicModulesUrl } from '../utils';
import { libraryVarName } from '../constants';

// eslint-disable-next-line import/prefer-default-export
export function createHotModuleWebpackConfig({
  context = getContextPath(),
  publicPath = getPublicModulesUrl(),
  staticPath = getModulesPath(),
  modules = [],
  externals = [],
  rootModuleName,
}) {
  return merge(
    externals.length > 0 ? createDLLConfig({
      externals: createOneAppExternals(externals),
      useAsReference: true,
    }) : {},
    {
      entry: createHotModuleEntries(modules),
      externals: Object.keys(createOneAppExternals(externals)),
      target: 'web',
      mode: 'development',
      devtool: 'source-map',
      context,
      output: {
        publicPath,
        path: staticPath,
        filename: '[name]/[name].js',
        hotUpdateChunkFilename: '[name]/[id].[fullhash].hot-update.js',
        uniqueName: libraryVarName,
      },
      module: {
        rules: [
          jsxLoader({ plugins: [require.resolve('react-refresh/babel')] }),
          cssLoader(),
          fileLoader(),
        ],
      },
      optimization: {
        runtimeChunk: 'single',
      },
      resolve: {
        mainFields: ['module', 'browser', 'main'],
        modules: [
          'node_modules',
          path.relative(context, path.resolve(__dirname, '..', '..', 'node_modules')),
        ].concat(modules.map(({ modulePath }) => path.join(modulePath, 'node_modules'))),
        extensions: ['.js', '.jsx'],
      },
      resolveLoader: {
        modules: [
          'node_modules',
          path.relative(context, path.resolve(__dirname, '..', '..', 'node_modules')),
        ].concat(modules.map(({ modulePath }) => path.join(modulePath, 'node_modules'))),
      },
      plugins: [
        new webpack.EnvironmentPlugin({
          // TODO: include environment variables being passed in from config
          NODE_ENV: 'development',
        }),
        new webpack.DefinePlugin({
          'global.BROWSER': JSON.stringify(true),
        }),
        new webpack.HotModuleReplacementPlugin(),
        // source injection
        new HotHolocronModulePlugin({
          rootModuleName,
          modules,
          externals,
        }),
        new ReactRefreshWebpackPlugin({
          library: libraryVarName,
          overlay: {
            sockIntegration: 'whm',
          },
        }),
      ],
    }
  );
}
