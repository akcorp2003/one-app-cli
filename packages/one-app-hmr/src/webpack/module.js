/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import webpack from 'webpack';
import merge from 'webpack-merge';
import CopyPlugin from 'copy-webpack-plugin';
import ExtractCssChunks from 'extract-css-chunks-webpack-plugin';
import ReactRefreshWebpackPlugin from '@pmmmwh/react-refresh-webpack-plugin';

import HotHolocronModulePlugin from './plugins/hot-holocron';

import {
  createOneAppExternals,
  createHotModuleEntries,
} from './utility';
import {
  jsxLoader,
  cssLoader,
  fileLoader,
} from './loaders';
import { createDLLConfig } from './dll';
import { debug } from '../logs';

// eslint-disable-next-line import/prefer-default-export
export function createHotModuleWebpackConfig({
  context,
  publicPath,
  staticPath,
  modules = [],
  externals = [],
  rootModuleName,
}) {
  const languagePacksToCopy = modules
    .filter(({ languagePacks }) => languagePacks.length > 0)
    .map(({ moduleName, moduleVersion, modulePath }) => ({
      from: path.join('**', `${moduleName}.json`),
      to: `${moduleName}/`,
      context: path.join(modulePath, 'build', moduleVersion),
    }));

  const entryConfig = {
    entry: createHotModuleEntries(modules),
  };

  debug('Webpack entries configuration used %o', entryConfig);

  return merge(
    entryConfig,
    (externals && externals.length > 0 ? createDLLConfig({
      dllAsReference: true,
    }) : {}),
    {
      externals: createOneAppExternals(externals),
      target: 'web',
      mode: 'development',
      devtool: 'source-map',
      context,
      output: {
        path: staticPath,
        filename: '[name]/[name].js',
        hotUpdateChunkFilename: '[name]/[id].[hash].hot-update.js',
        publicPath: `/${publicPath}/`,
      },
      resolve: {
        mainFields: ['module', 'browser', 'main'],
        modules: [
          'node_modules',
          path.relative(context, path.resolve(__dirname, '..', '..', 'node_modules')),
        ].concat(modules.map(({ modulePath }) => path.join(modulePath, 'node_modules'))),
        extensions: ['.js', '.jsx'],
      },
      resolveLoader: {
        modules: [
          'node_modules',
          path.resolve(__dirname, '..', '..', 'node_modules'),
        ],
        extensions: ['.js', '.json'],
        mainFields: ['loader', 'main'],
      },
      module: {
        rules: [
          jsxLoader({ plugins: ['react-refresh/babel'] }),
          cssLoader,
          fileLoader,
        ],
      },
      plugins: [
        new webpack.EnvironmentPlugin({
          NODE_ENV: 'development',
        }),
        new webpack.DefinePlugin({
          'global.BROWSER': JSON.stringify(true),
        }),
        new webpack.HotModuleReplacementPlugin(),
        new ReactRefreshWebpackPlugin({
          overlay: {
            sockIntegration: 'whm',
          },
        }),
        // source injection
        new HotHolocronModulePlugin({
          rootModuleName,
          modules,
          externals,
        }),
        // css/styles
        new ExtractCssChunks({
          filename: '[name]/[name].css',
        }),
      ].concat(
        languagePacksToCopy.length > 0
        // copies (and updates) the language packs for a module
          ? new CopyPlugin({
            patterns: languagePacksToCopy,
          })
          : []
      ),
    }
  );
}
