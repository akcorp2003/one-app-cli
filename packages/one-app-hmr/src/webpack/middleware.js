/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import webpack from 'webpack';
import webpackDevMiddleware from 'webpack-dev-middleware';
import webpackHotMiddleware from 'webpack-hot-middleware';

import {
  createOneAppExternals,
} from './utility';
import { createDLLConfig } from './dll';
import { createHotModuleWebpackConfig } from './module';
import {
  error, warn, time, log, orange, dodgerblue,
} from '../logs';

export const printWebpack = (message) => `${dodgerblue('webpack')} - ${message}`;

export function buildExternalsDLL(config = {}) {
  const { externals = [] } = config;

  if (!(externals.length > 0)) return Promise.resolve();

  return new Promise((resolve, reject) => {
    webpack(createDLLConfig({
      isDev: false,
      dllVendors: externals,
      dllExternals: createOneAppExternals(),
    })).run((err, stats) => {
      if (err) {
        error(err);
        reject(err);
      }
      resolve(stats);
    });
  });
}

export async function loadWebpackMiddleware({
  context,
  publicPath,
  staticPath,
  modules,
  externals,
  rootModuleName,
} = {}) {
  log(printWebpack('initializing webpack'));

  await time(printWebpack(orange('pre-building dll externals for Holocron modules')), async () => {
    await buildExternalsDLL({ externals });
  });

  // TODO: consider building externals if changed
  const webpackConfig = createHotModuleWebpackConfig({
    context,
    publicPath,
    staticPath,
    modules,
    externals,
    rootModuleName,
  });
  const compiler = webpack(webpackConfig);
  const devMiddleware = webpackDevMiddleware(compiler, {
    index: false,
    serverSideRender: true,
    writeToDisk: true,
    logTime: false,
    logLevel: 'error',
    publicPath: webpackConfig.output.publicPath,
    watchOptions: {
      aggregateTimeout: 500,
      poll: 1000,
    },
    reporter: (opts, { state, stats } = {}) => {
      if (state && stats) {
        const { errors, warnings } = stats.compilation;
        errors.forEach((message) => error(message));
        warnings.forEach((message) => warn(message));
        log(printWebpack(`webpack built in ${orange(`${stats.endTime - stats.startTime}`)} ms`));
      } else {
        warn(printWebpack('webpack building...'));
      }
      return null;
    },
  });
  const hotMiddleware = webpackHotMiddleware(compiler, {
    reload: false,
    log: false,
    dynamicPublicPath: false,
  });

  return {
    devMiddleware,
    hotMiddleware,
    publish: (...args) => hotMiddleware.publish(...args),
  };
}
