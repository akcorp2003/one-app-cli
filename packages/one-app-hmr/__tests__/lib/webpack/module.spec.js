/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { validate } from 'webpack';
import { createHotModuleWebpackConfig } from '../../../lib/webpack/module';

describe('createHotModuleWebpackConfig', () => {
  test('creates the webpack config for hot reloadable modules', () => {
    const modules = [{
      moduleName: 'hot-module',
      modulePath: 'hot-module/src/index.js',
    }];
    const config = createHotModuleWebpackConfig({
      modules,
    });
    expect(validate(config)).toBe(undefined);
  });

  test('uses DLL config when externals are provided', () => {
    const externals = ['react-package'];
    const config = createHotModuleWebpackConfig({
      externals,
    });
    expect(validate(config)).toBe(undefined);
    expect(config.externals).toEqual([
      '@americanexpress/one-app-ducks',
      '@americanexpress/one-app-router',
      'create-shared-react-context',
      'holocron',
      'holocron-module-route',
      'immutable',
      'prop-types',
      'react',
      'react-dom',
      'react-helmet',
      'react-redux',
      'redux',
      'reselect',
      // ensure provided/required externals are added to the main set of one app externals
      'react-package',
    ]);
  });
});
