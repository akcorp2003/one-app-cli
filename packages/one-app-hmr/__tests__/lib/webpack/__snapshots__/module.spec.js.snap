// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHotModuleWebpackConfig returns the optimization webpack config for minifying the output 1`] = `
Object {
  "context": "/Users/mmtobia/code/one-app-cli",
  "devtool": "source-map",
  "entry": Object {},
  "externals": Array [
    "@americanexpress/one-app-ducks",
    "@americanexpress/one-app-router",
    "create-shared-react-context",
    "holocron",
    "holocron-module-route",
    "immutable",
    "prop-types",
    "react",
    "react-dom",
    "react-helmet",
    "react-redux",
    "redux",
    "reselect",
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "/Users/mmtobia/code/one-app-cli/node_modules/babel-loader/lib/index.js",
            "options": Object {
              "babelrc": false,
              "cacheDirectory": true,
              "cwd": "/Users/mmtobia/code/one-app-cli/packages/one-app-hmr/lib",
              "plugins": Array [
                "/Users/mmtobia/code/one-app-cli/node_modules/react-refresh/babel.js",
              ],
              "presets": Array [
                Array [
                  "amex",
                  Object {
                    "modern": true,
                    "preset-env": Object {
                      "modules": false,
                    },
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
        "use": Array [
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "importLoaders": 2,
              "modules": Object {
                "localIdentName": "[name]__[local]___[contenthash:base64:5]",
              },
            },
          },
          Object {
            "loader": "sass-loader",
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Object {
                  "cssnano": Object {},
                  "postcss-browser-reporter": Object {},
                  "postcss-preset-env": Object {
                    "browsers": "last 2 versions",
                  },
                },
                "syntax": "postcss-scss",
              },
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(woff\\|woff2\\|ttf\\|eot\\|svg\\|png\\|jpg\\|jpeg\\|gif\\|webm\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": "../assets/[name].[ext]",
            },
          },
        ],
      },
    ],
  },
  "optimization": Object {
    "runtimeChunk": "single",
  },
  "output": Object {
    "filename": "[name]/[name].js",
    "hotUpdateChunkFilename": "[name]/[id].[fullhash].hot-update.js",
    "path": "/Users/mmtobia/code/one-app-cli/static/modules",
    "publicPath": "/static/modules/",
    "uniqueName": "oneAppHmr",
  },
  "plugins": Array [
    EnvironmentPlugin {
      "defaultValues": Object {
        "NODE_ENV": "development",
      },
      "keys": Array [
        "NODE_ENV",
      ],
    },
    DefinePlugin {
      "definitions": Object {
        "global.BROWSER": "true",
      },
    },
    HotModuleReplacementPlugin {
      "options": Object {},
    },
    HotHolocronModulePlugin {
      "options": Object {
        "externals": Array [],
        "modules": Array [],
        "rootModuleName": undefined,
      },
    },
    ReactRefreshPlugin {
      "options": Object {
        "exclude": /node_modules/i,
        "include": /\\\\\\.\\(\\[jt\\]sx\\?\\|flow\\)\\$/i,
        "library": "oneAppHmr",
        "overlay": Object {
          "entry": "/Users/mmtobia/code/one-app-cli/packages/one-app-hmr/node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js",
          "module": "/Users/mmtobia/code/one-app-cli/packages/one-app-hmr/node_modules/@pmmmwh/react-refresh-webpack-plugin/overlay/index.js",
          "sockIntegration": "whm",
        },
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
    ],
    "mainFields": Array [
      "module",
      "browser",
      "main",
    ],
    "modules": Array [
      "node_modules",
      "packages/one-app-hmr/node_modules",
    ],
  },
  "resolveLoader": Object {
    "modules": Array [
      "node_modules",
      "packages/one-app-hmr/node_modules",
    ],
  },
  "target": "web",
}
`;
